src/BizHawk.Emulation.Cores/Consoles/3DO/LibOpera.cs

```diff
@@ -0,0 +1,159 @@
+﻿using System.Runtime.InteropServices;
+using BizHawk.BizInvoke;
+using BizHawk.Emulation.Cores.Waterbox;
+
+namespace BizHawk.Emulation.Consoles._3DO
+{
+	public abstract class LibOpera : LibWaterboxCore
+	{
+		// NTSC Specifications
+		public const int NTSC_WIDTH = 320;
+		public const int NTSC_HEIGHT = 240;
+		public const int NTSC_VIDEO_NUMERATOR = 60;
+		public const int NTSC_VIDEO_DENOMINATOR = 1;
+
+		// PAL1 Specifications
+		public const int PAL1_WIDTH = 320;
+		public const int PAL1_HEIGHT = 288;
+		public const int PAL1_VIDEO_NUMERATOR = 50;
+		public const int PAL1_VIDEO_DENOMINATOR = 1;
+
+		// PAL2 Specifications
+		public const int PAL2_WIDTH = 384;
+		public const int PAL2_HEIGHT = 288;
+		public const int PAL2_VIDEO_NUMERATOR = 50;
+		public const int PAL2_VIDEO_DENOMINATOR = 1;
+
+		[UnmanagedFunctionPointer(CC)]
+		public delegate void CDReadCallback(int lba, IntPtr dst);
+
+		[UnmanagedFunctionPointer(CC)]
+		public delegate int CDSectorCountCallback();
+
+		[BizImport(CC)]
+		public abstract void SetCdCallbacks(CDReadCallback cdrc, CDSectorCountCallback cdscc);
+
+        [BizImport(CC)]
+        public abstract void ejectCD();
+
+        [BizImport(CC)]
+        public abstract void insertCD();
+
+        [BizImport(CC, Compatibility = true)]
+		public abstract bool Init(string gameFile, string biosFile, string fontFile, int port1Type, int port2Type, int videoStandard);
+
+		[BizImport(CC, Compatibility = true)]
+		public abstract bool sram_changed();
+
+		[BizImport(CC, Compatibility = true)]
+		public abstract int get_sram_size();
+
+		[BizImport(CC, Compatibility = true)]
+		public abstract void get_sram(IntPtr sramBuffer);
+
+		[BizImport(CC, Compatibility = true)]
+		public abstract void set_sram(IntPtr sramBuffer);
+
+		[StructLayout(LayoutKind.Sequential)]
+		public struct GamepadInputs
+		{
+			public int up;
+			public int down;
+			public int left;
+			public int right;
+			public int start;
+			public int select;
+			public int buttonA;
+			public int buttonB;
+			public int buttonX;
+			public int buttonY;
+			public int buttonL;
+			public int buttonR;
+		}

```

Comment:
Is there a reason these are all `int`s instead of a type more closely resembling their data range, like `bool`?

Reply:
I did try using `bool` but I keep getting false at the core side, even when I send trues from the bk side. I looked high and low for a discrepancy in the apis, but everything looks correct.

Reply:
When using bool, was the core actually bool too, or just ints? The struct sizes need to match up.

Also the `Compatibility = true` would also probably cause issues regardless if you did that, as that would go under "normal" marshalling rules (i.e. the struct is converted into one with 4 byte bools instead of just directly pinning the original struct). You generally shouldn't use Compatibility = true unless you need "normal" marshalling as that slower.

Reply:
I tried again, making sure again that the structs coincide. You can find them here: 

[BK Side]
https://github.com/SergioMartin86/BizHawk/blob/b4836176a8df5d2eba2d77598412036767840d8d/src/BizHawk.Emulation.Cores/Consoles/3DO/LibOpera.cs#L58

[WBX side] https://github.com/SergioMartin86/BizHawk/blob/b4836176a8df5d2eba2d77598412036767840d8d/waterbox/opera/bizhawk.hpp#L13

The values are simply not passed around:
![image](https://github.com/user-attachments/assets/b1fc3164-164a-443d-9ed3-fc570256ac0a)

Perhaps you will see the discrepancy -- I just don't see it

Reply:
It's probably got something to do with the bullshit default bool marshalling rules that nobody wants, although `BizInvoke` should usually get rid of that...

Reply:
In the case Compatibility = false, it does get "rid" of it, as Compatibility = false always just pins the struct and passes the pointer along (never any conversion, great as long as you only have blittable members). In the case of Compatibility = true, it has to go through standard marshalling code as you're just calling a delegate from GetDelegateForFunctionPointer (which is again much slower than using Compatibility = false)

Reply:
Yeah but there is no compatibility in the usage of this struct (just the frame advance call as far as I can see)

Reply:
Ah, in that case it's getting placed into a class, which has some strange behavior in this regard, since this isn't strictly blittable behavior overall. For saving space and just using bools, `byte` would be used C# side (or maybe rather bit flags if ypu really want to conserve space, which might or might not be wanted since bigger frame info means more GC pressure)

Reply:
Adding a `uint64_t xdd;` at the start of `struct gamePad_t` makes it work, so apparently there's 8 extra padding bytes in the c# class somehow. There more I've read into marshalling and blitting the more confused I get as to how things even work in other cores. It's possible everything that is working is just working by pure chance.

Reply:
Could be weird mismatch of C# vs C padding rules? In the perspective of marshalling, a lot of usual knowledge doesn't *strictly* apply per se, since the class instance just gets pinned down and the pointer to such is passed on, so it matters how C# is internally ordering the data.

Reply:
What's the state of this discussion?

Reply:
I think the problem is that bool types don't count as blittable types for whatever garbage legacy reason in C# which is most likely the reason using bool in any struct at all no longer guarantees struct layout. So as unsatisfied as I am with the conclusion I say leave it as is.

Reply:
I don't think so. Applying atrributes to every bool would only change the layout in interop which we're not using here.